shared/js/toc.jsx

```diff
@@ -0,0 +1,772 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+var loop = loop || {};
+loop.shared = loop.shared || {};
+
+loop.shared.toc = (function(mozL10n) {
+  "use strict";
+
+  var ROOM_INFO_FAILURES = loop.shared.utils.ROOM_INFO_FAILURES;
+  var ROOM_STATES = loop.store.ROOM_STATES;
+  var sharedActions = loop.shared.actions;
+  var sharedMixins = loop.shared.mixins;
+  var sharedViews = loop.shared.views;
+
+  var TableOfContentView = React.createClass({
+    propTypes: {
+      activeRoomStore: React.PropTypes.instanceOf(loop.store.ActiveRoomStore).isRequired,
+      dispatcher: React.PropTypes.instanceOf(loop.Dispatcher).isRequired,
+      isDesktop: React.PropTypes.bool.isRequired,
+      isScreenShareActive: React.PropTypes.bool.isRequired
+    },
+
+    getInitialState: function() {
+      return this.props.activeRoomStore.getStoreState();
+    },
+
+    componentWillMount: function() {
+      this.props.activeRoomStore.on("change", this.onStoreChange);
+      // Force onStoreChange
+      this.onStoreChange();
+    },
+
+    componentWillUnmount: function() {
+      this.props.activeRoomStore.off("change", this.onStoreChange);
+    },
+
+    componentDidMount: function() {
+      if (this.props.isDesktop) {
+        document.body.classList.add("desktop-toc");
+      }
+    },
+
+    onStoreChange: function() {
+      var newState = this.props.activeRoomStore.getStoreState();
+      var tiles = newState.roomContextUrls ? [newState.roomContextUrls[0]] : [];
+      newState.tiles = tiles;
```

Comment:
This feels like a workaround for a bad default store state. Really the store should be setting the right information here.